Storage/SQL/SQLiteHistoryRecommendations.swift

```diff
@@ -50,60 +50,29 @@ extension SQLiteHistory: HistoryRecommendations {
         return self.db.runQuery(sql, args: nil, factory: SQLiteHistory.iconHistoryMetadataColumnFactory)
     }
 
-    public func invalidateHighlights() -> Success {
-        return clearHighlights() >>> populateHighlights
-    }
-
     public func removeHighlightForURL(_ url: String) -> Success {
         return self.db.run([("INSERT INTO \(TableActivityStreamBlocklist) (url) VALUES (?)", [url])])
     }
 
-    public func clearHighlights() -> Success {
-        return self.db.run("DELETE FROM \(TableHighlights)", withArgs: nil)
-    }
-
-    private func populateHighlights() -> Success {
+    private func repopulateHighlightsQuery() -> [(String, Args?)] {
         let (query, args) = computeHighlightsQuery()
+        let clearHighlightsQuery = "DELETE FROM \(TableHighlights)"
 
-        // Convert the fetched row into arguments for a bulk insert along with the
-        // generated cache_key value.
-        func argsFrom(row: SDRow) -> Args? {
-            let urlString = row["url"] as! String
-            let cacheKey = SQLiteMetadata.cacheKeyForURL(urlString.asURL!)!
-            return [
-                row["historyID"],
-                cacheKey,
-                urlString,
-                row["title"],
-                row["guid"],
-                row["visitCount"],
-                row["visitDate"],
-                row["is_bookmarked"]
-            ]
+        let sql = "INSERT INTO \(TableHighlights) " +
+            "SELECT historyID, url as cache_key, url, title, guid, visitCount, visitDate, is_bookmarked " +
+            "FROM (\(query))"
+        return [(clearHighlightsQuery, nil), (sql, args)]
+    }
+
+    public func repopulateAll(_ invalidateTopSites: Bool, invalidateHighlights: Bool) -> Success {
+        var queries: [(String, Args?)] = []
+        if invalidateHighlights {
+            queries.append(contentsOf: self.repopulateHighlightsQuery())
```

Comment:
… cos I'd hate for us to do a bunch of unnecessary string interpolation and allocation here each time, just to run the same fixed SQL.